walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,105 @@
+          return EMPTY_LIST;
+      }
+      if(eventsList.size()==0){
+          return asList(TimeRange.WHOLE_DAY);
+      }
+      if(eventsList.size()==1){
+          return asList(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, eventsList.get(0).getWhen().start(), false),
+          TimeRange.fromStartEnd(eventsList.get(0).getWhen().end(), TimeRange.END_OF_DAY, true)); 
+       }
+
+      return getOpenTimeSlots(eventsList, request);
+  }
+  /**
+  *  This method gets the open time slots by looping through a list of events
+  *  given that the list is sorted by start time. 
+  *  @param eventsList list of events in sorted order (by start time).
+  *  @param request he duration of requested meeting and people attending.
+  *  @return A list of open meeting timeslots. 
+  */
+  private Collection<TimeRange> getOpenTimeSlots(List<Event> eventsList, MeetingRequest request){
+      Collection<TimeRange> queryResult = new ArrayList<>();
+
+      Optional<TimeRange> optionalTimeRange;
+
+      optionalTimeRange = hasSufficientTime(TimeRange.START_OF_DAY, eventsList.get(0).getWhen().start(), request.getDuration(), false);
+      optionalTimeRange.ifPresent(range -> queryResult.add(range));
+      
+      for(int i=0;i<eventsList.size();i++){
+        // First event in the list, check if there is enough time between event start and the start of the day. 
+        //  If there is enough time, add to the result. Since the events are sorted by start time we can check the
+        //  first index.
+        TimeRange eventTime = eventsList.get(i).getWhen();
+        int eventEnd = eventTime.end();
+
+        // If this isn't the last event.
+        if(i != eventsList.size()-1){
+          TimeRange nextEvent = eventsList.get(i+1).getWhen();
+          if(eventTime.overlaps(nextEvent)){
+            // Check for overlap.  Don't need to check for the reverse, since it is pre-sorted. 
+            if(eventTime.contains(nextEvent)){
+              //  If one event contains another event. Set i+1 to i.  
+              // This is to get the earliest start and latest end time for an event.
+              eventsList.set(i+1, eventsList.get(i));
+            }   
+        //  Check if there is enough time between two events. 
+        } else {
+           optionalTimeRange = hasSufficientTime(eventEnd, nextEvent.start(), 
+             request.getDuration(), false);
+           optionalTimeRange.ifPresent(range -> queryResult.add(range));
+         } 
+        // For last element in event list check if there is enough time between event end and end of day 
+       } else {
+          optionalTimeRange = hasSufficientTime(eventEnd,TimeRange.END_OF_DAY, 
+            request.getDuration(), true);
+          optionalTimeRange.ifPresent(range -> queryResult.add(range));
+       }
+      }
+      return queryResult;
+  }
+
+  private Optional<TimeRange> hasSufficientTime(int start, int end, long duration, boolean inclusive){
```

Comment:
is/has/was are reserved for boolean methods.  getTimeSlotWhenPossible

Reply:
Fixed.